File: bench/diff_bcd.c
// SPDX-License-Identifier: GPL-2.0
// Tier-2.5 differential oracle: C original vs Rust translation, bcd.
// Reference extracted from lib/bcd.c (v7.1); kept byte-identical.
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>

static unsigned _bcd2bin(unsigned char val)
{
	return (val & 0x0f) + (val >> 4) * 10;
}

static unsigned char _bin2bcd(unsigned val)
{
	const unsigned int t = (val * 103) >> 10;

	return (t << 4) | (val - t * 10);
}

// Explicit LCG (same constants used across all bench/diff_*.c files).
static uint64_t lcg_state;
static uint32_t lcg_next(void)
{
	lcg_state = lcg_state * 6364136223846793005ULL + 1442695040888963407ULL;
	return (uint32_t)(lcg_state >> 32);
}

int main(int argc, char **argv)
{
	long n = argc > 1 ? atol(argv[1]) : 5000;
	lcg_state = argc > 2 ? (uint64_t)atol(argv[2]) : 271828;

	for (long i = 0; i < n; i++) {
		unsigned char val = (unsigned char)lcg_next();
		unsigned r = _bcd2bin(val);
		printf("bcd2bin,%u,%u\n", val, r);
	}

	for (long i = 0; i < n; i++) {
		// _bin2bcd's contract is "val < 100" (bin2bcd(val), val is a BCD-
		// representable binary value); exercise that domain plus a few
		// values just outside it, matching what real kernel callers pass.
		unsigned val = lcg_next() % 100;
		unsigned char r = _bin2bcd(val);
		printf("bin2bcd,%u,%u\n", val, r);
	}
	return 0;
}

File: bench/diff_bcd.rs
// SPDX-License-Identifier: GPL-2.0
//! Tier-2.5 differential oracle: Rust translation side, bcd. Faithful
//! copy of lib/bcd_rs.rs's algorithm (kernel-crate bindings/export
//! stripped) — same protocol/LCG as diff_bcd.c.

fn bcd2bin(val: u8) -> u32 {
    ((val & 0x0f) as u32) + ((val >> 4) as u32) * 10
}

fn bin2bcd(val: u32) -> u8 {
    let t = val.wrapping_mul(103) >> 10;
    ((t << 4) | (val - t * 10)) as u8
}

// Identical LCG to diff_bcd.c.
struct Lcg(u64);
impl Lcg {
    fn next(&mut self) -> u32 {
        self.0 = self
            .0
            .wrapping_mul(6364136223846793005)
            .wrapping_add(1442695040888963407);
        (self.0 >> 32) as u32
    }
}

fn main() {
    let args: Vec<String> = std::env::args().collect();
    let n: u64 = args.get(1).and_then(|a| a.parse().ok()).unwrap_or(5000);
    let seed: u64 = args.get(2).and_then(|a| a.parse().ok()).unwrap_or(271828);
    let mut rng = Lcg(seed);

    for _ in 0..n {
        let val = (rng.next() & 0xff) as u8;
        let r = bcd2bin(val);
        println!("bcd2bin,{},{}", val, r);
    }

    for _ in 0..n {
        let val = rng.next() % 100;
        let r = bin2bcd(val);
        println!("bin2bcd,{},{}", val, r);
    }
}
